amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -116,24 +153,224 @@ public void stop() {
+                    }
+                }
+                return lease;
+            }).filter(lease -> lease.hashKeyRangeForLease() != null).collect(Collectors.toList());
+        } else {
+            return activeLeases;
+        }
+    }
+
+    @VisibleForTesting
+    static Optional<HashRangeHole> checkForHoleInHashKeyRanges(StreamIdentifier streamIdentifier,
+            List<HashKeyRangeForLease> hashKeyRanges, BigInteger minHashKey, BigInteger maxHashKey) {
+        // Sort and merge the overlapping hash ranges.
+        List<HashKeyRangeForLease> mergedHashKeyRanges = sortAndMergeOverlappingHashRanges(hashKeyRanges);
+
+        if(mergedHashKeyRanges.isEmpty()) {
+            log.error("No valid hashranges found for stream {} between {} and {}.", streamIdentifier,
+                    MIN_HASH_KEY, MAX_HASH_KEY);
+            return Optional.of(new HashRangeHole(new HashKeyRangeForLease(MIN_HASH_KEY, MAX_HASH_KEY),
+                    new HashKeyRangeForLease(MIN_HASH_KEY, MAX_HASH_KEY)));
+        }
+
+        // Validate for hashranges bounds.
+        if (!mergedHashKeyRanges.get(0).startingHashKey().equals(minHashKey) || !mergedHashKeyRanges
+                .get(mergedHashKeyRanges.size() - 1).endingHashKey().equals(maxHashKey)) {
+            log.error("Incomplete hash range found for stream {} between {} and {}.", streamIdentifier,
+                    mergedHashKeyRanges.get(0),
+                    mergedHashKeyRanges.get(mergedHashKeyRanges.size() - 1));
+            return Optional.of(new HashRangeHole(mergedHashKeyRanges.get(0),
+                    mergedHashKeyRanges.get(mergedHashKeyRanges.size() - 1)));
+        }
+        // Check for any holes in the sorted hashrange intervals.
+        if (mergedHashKeyRanges.size() > 1) {
+            for (int i = 1; i < mergedHashKeyRanges.size(); i++) {
+                final HashKeyRangeForLease hashRangeAtStartOfPossibleHole = mergedHashKeyRanges.get(i - 1);
+                final HashKeyRangeForLease hashRangeAtEndOfPossibleHole = mergedHashKeyRanges.get(i);
+                final BigInteger startOfPossibleHole = hashRangeAtStartOfPossibleHole.endingHashKey();
+                final BigInteger endOfPossibleHole = hashRangeAtEndOfPossibleHole.startingHashKey();
+
+                if (!endOfPossibleHole.subtract(startOfPossibleHole).equals(BigInteger.ONE)) {
+                    log.error("Incomplete hash range found for {} between {} and {}.", streamIdentifier,
+                            hashRangeAtStartOfPossibleHole, hashRangeAtEndOfPossibleHole);
+                    return Optional.of(new HashRangeHole(hashRangeAtStartOfPossibleHole, hashRangeAtEndOfPossibleHole));
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    @VisibleForTesting
+    static List<HashKeyRangeForLease> sortAndMergeOverlappingHashRanges(
+            List<HashKeyRangeForLease> hashKeyRanges) {
+        if (hashKeyRanges.size() == 0 || hashKeyRanges.size() == 1)
```

Comment:
This method is not required. We can iterate over the lease with hash range min and verify it starts from 0 to the max (2^128-1)

If lease has overlap and hash range start and end is less than the current max. We can safely move to the next item and validate